### PR TITLE
[dev] moe(perf): Refactor CP layout organization for Qwen3.5-style hybrid attention model

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

PR for main: https://github.com/NVIDIA/Megatron-LM/pull/6515

This PR adds fixed context-parallel (CP) layout support for long-context hybrid models that want to run most layers in `cp_partition_mode="contiguous"` while still keeping regular attention/GQA in the `zigzag` layout it requires.

image

The main goals are:

- Move CP layout conversion to the GQA/regular-attention boundary for Qwen3.5-style schedules. In the target GQA:GDN = 1:3 pattern, GDN chunkwise CP can consume contiguous layout directly, so only the less frequent GQA layers need to convert to zigzag and back. This reduces layout conversion frequency by roughly two thirds.
- Prebuild THD CP layout routes once per microbatch. THD layout conversion can then reuse precomputed split/routing metadata instead of rebuilding the route on the conversion hot path.

E2E perf gain (A proxy model with the same single-layer size as Qwen3.5-397B):

- Prebuild THD CP layout routes: 1.20x speedup
- Globally using contiguous layout: 1.075x speedup
- Overall: 1.29x speedup

## Motivation

The previous fixed CP layout path effectively kept the model in zigzag layout because regular attention requires zigzag. That is correct for attention, but it is not optimal for Qwen3.5-style hybrid stacks where most layers are GDN layers.

For chunkwise CP, GDN can run in contiguous layout. This PR lets the model keep the broader stack in contiguous layout, then converts only when entering regular attention/GQA and converts the attention output back afterward. With a GQA:GDN = 1:3 schedule, this shifts layout shuffles away from the dense GDN path and places them on the sparser attention boundary.

For THD packed sequences, the route for a given microbatch is deterministic. This PR moves route construction into batch preparation and reuses the result for later `zigzag <-> contiguous` conversions.

### 1. CP layout conversion package

- Splits `megatron/core/context_parallel_layout.py` into a package:
  - `conversion.py`: SBHD/THD tensor conversion and `CpPartitionModeConverter`.
  - `routes.py`: THD route construction, encoding/decoding, lookup, and prebuild.
  - `metadata.py`: helpers for packed sequence layout metadata.
  - `policy.py`: stage-entry layout policy helpers.
  - `__init__.py`: compatibility exports for existing imports.
- Adds `convert_module_input_tensors_cp_partition_mode()` for module-boundary conversion:
  - converts module inputs to the required concrete layout;
  - returns a converter for mapping module outputs back to the caller layout;
  - rejects unsupported dense attention masks/bias and cross-attention key/value states.

### 2. Attention/GDN layout boundary

- Regular attention converts its input to zigzag when the caller layout is different, then converts the output back to the caller layout.
- GDN chunkwise CP can run directly in contiguous layout.
- GDN headwise CP keeps the existing zigzag requirement.
- This makes `cp_partition_mode="contiguous"` useful for Qwen3.5-style schedules: most GDN layers avoid conversion, while GQA/regular-attention layers handle their own layout boundary.

### 3. THD route prebuild

- Adds `prebuild_thd_cp_partition_routes()` to compute both THD conversion directions during batch construction.
- Updates GPT and Hybrid `get_batch` paths to annotate the stage-entry layout and prebuild THD routes before model forward.
- Updates route lookup so conversion can consume the prebuilt route. Missing routes still fall back to lazy construction with a warning for compatibility, but the intended path is prebuild during batch preparation.

### 4. Checks and unsupported cases

This PR adds two layers of validation.

First, `TransformerConfig` rejects unsupported configuration combinations early:

- `contiguous` CP currently requires `experimental_attention_variant` to be `gated_delta_net` or `dsv4_hybrid`.
- `dsv4_hybrid` with CP requires `contiguous`.
- `gated_delta_net` with `linear_cp_mode="headwise"` cannot use `contiguous`.
- `multi_latent_attention` outside `dsv4_hybrid` cannot use `contiguous`.

Second, runtime layout is carried through `PackedSeqParams.cp_partition_mode`, and attention variants validate the layout they receive:

- Regular attention requires zigzag.
- GDN chunkwise CP requires contiguous.
- GDN headwise CP requires zigzag.
- MTP validates its rolling path and rejects non-packed/SBHD contiguous layout until that path is implemented.

These checks are meant to fail early when tensor order would otherwise silently diverge from the module's layout assumption.

## Limitations / Future Work

- Dense attention masks, attention bias, and cross-attention key/value states are still rejected for CP layout conversion.
- Non-packed/SBHD contiguous MTP is intentionally not supported yet.
- Full-iteration CUDA graph is not supported for THD CP layout conversion today because torch-native NCCL A2A only supports fixed splits. Supporting this path will require GIN, and we plan to enable it through DeepEP / HybridEP dispatch.
- TP x CP conversion can be further optimized to use a single A2A communication. The current implementation still needs two communication steps.
- Auto conversion follow-up: <https://github.com/NVIDIA/Megatron-LM/pull/5664|[dev] moe(perf): Refactor CP layout - auto layout conversion by yuzhongw-nvidia · Pull Request #5664 · NVIDIA/Megatron-LM>

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/ssm/gated_delta_net.py [hunk 1]
```diff
@@ -295,6 +296,12 @@ def reset_parameters(self):
                 # conv1d.weight
                 if self.conv_init is not None:
                     nn.init.uniform_(self.conv1d.weight, -self.conv_init, self.conv_init)
+                else:
+                    nn.init.kaiming_uniform_(self.conv1d.weight, a=math.sqrt(5))
+                if self.conv1d.bias is not None:
+                    fan_in = self.conv1d.weight.size(1) * self.conv1d.weight.size(2)
+                    bound = 1 / math.sqrt(fan_in)
+                    nn.init.uniform_(self.conv1d.bias, -bound, bound)
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py]: Will remove before merge. Thanks.

File: tests/unit_tests/ssm/test_gdn_moe_cp_loss_parity.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+except ImportError:
+    HAVE_FLA = False
+
+
+_PARALLEL_EP_SIZE = 4
+_SEQUENCE_LENGTH = 4096
+_MICRO_BATCH_SIZE = 1
+_VOCAB_SIZE = 8192
+_SEED = 1234
+_DIAGNOSTIC_REPEATS = 5
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply tests/unit_tests/ssm/test_gdn_moe_cp_loss_parity.py]: Will remove before merge. Thanks.

File: megatron/core/context_parallel_layout/__init__.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+- model postprocess restores the public output boundary to the input layout;
+- MTP validates its inner-layer layout preference but does not own outer conversion.
+"""
+
+from typing import Literal
+
+CpPartitionMode = Literal["zigzag", "contiguous"]
+
+from megatron.core.context_parallel_layout.conversion import (
+    CpPartitionModeConverter,
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout/__init__.py]: Reject. Thanks.

File: megatron/core/context_parallel_layout/routes.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    chunk_divisor = 2 * cp_size
+    bad_seq_lens = [
+        seq_end - seq_start
+        for seq_start, seq_end in zip(cu[:-1], cu[1:])
+        if (seq_end - seq_start) % chunk_divisor != 0
+    ]
+    if bad_seq_lens:
+        raise ValueError(
+            "All packed sequence lengths must be divisible by "
+            f"2 * cp_size ({chunk_divisor}) for zigzag CP layout conversion, "
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout/routes.py:58]: Reject. Thanks.

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/8d8189cd829af1ffed4f8adf16488e31ba0dd336

File: megatron/core/transformer/transformer_config.py [hunk 5]
```diff
@@ -1566,23 +1566,62 @@ def __post_init__(self):
             if self.sequence_packing_scheduler is None:
                 raise ValueError(
                     "cp_partition_mode='contiguous' with context parallelism requires THD "
-                    "inputs from a sequence_packing_scheduler; BSHD inputs are not supported."
+                    "inputs from a sequence_packing_scheduler; BSHD inputs are not supported. "
+                    "The legacy non-scheduler CP batch slicing path only supports zigzag layout."
+                )
+            elif self.cuda_graph_impl == "full_iteration":
```
Comment (BestJuly) [inline megatron/core/transformer/transformer_config.py]: Does it mean current layout conversion is compatible with other CUDA graph impl? E.g., partial CUDA graph which include the `attn` part, as the conversion is in its the forward function.

File: megatron/core/transformer/attention.py [hunk 6]
```diff
@@ -1361,6 +1362,33 @@ def forward(
...
+            config=self.config,
+            attention_mask=attention_mask,
+            attention_bias=attention_bias,
+        )
+        if (
+            packed_seq_params is not None
+            and packed_seq_params.qkv_format == "thd"
+            and self.pg_collection.cp is not None
+            and get_pg_size(self.pg_collection.cp) > 1
+            and packed_seq_params.cp_partition_mode != "zigzag"
```
Comment (BestJuly) [inline megatron/core/transformer/attention.py]: In the process of `convert_module_input_tensors_cp_partition_mode`, the `packed_seq_params.cp_partition_mode` has already been modified to `zigzag`, so this condition seems to be useless.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 7]
```diff
@@ -909,13 +909,24 @@ def forward(
...
+        # Set the right cp group for dynamic-cp. Downstream RoPE and CSA core
+        # attention use self.pg_collection.cp, which must point at this
+        # microbatch's dynamic CP group. Restored before returning.
         _orig_cp_group = self.pg_collection.cp
         if packed_seq_params is not None and packed_seq_params.local_cp_size is not None:
             assert packed_seq_params.cp_group is not None, "cp_group must be set in dynamic-cp mode"
             self.pg_collection.cp = packed_seq_params.cp_group
+        if (
+            packed_seq_params is not None
+            and packed_seq_params.qkv_format == "thd"
```
Comment (BestJuly) [inline megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: nit: these two condition has been combined with `packed_seq` in this file 
```
packed_seq = packed_seq_params is not None and packed_seq_params.qkv_format == 'thd'
```

Also, just found in [mla](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/5709c670d7eddc52b25681c91383ff9bd7783233/megatron/core/transformer/multi_latent_attention.py#L398), `thd_packed_seq` is used. It is better to make it consistent.

File: megatron/core/context_parallel_layout/routes.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    get_packed_seq_params_cp_partition_cu_seqlens,
+)
+from megatron.core.context_parallel_layout.types import CpPartitionMode, ThdCpRoute
+
+if TYPE_CHECKING:
+    from megatron.core.packed_seq_params import PackedSeqParams
+
+
+@contextmanager
+def _cp_layout_nvtx_range(message: str):
```
Comment (BestJuly) [inline megatron/core/context_parallel_layout/routes.py]: There are already several types to add nvtx range in MCore and here is a new one. Could we consider to reduce the various usages such as reuse the one [here](https://github.com/NVIDIA/Megatron-LM/blob/574e04ebf802b3b3cfc90d97dcbfc0721e7313af/megatron/training/utils/common_utils.py#L627)

File: megatron/core/context_parallel_layout/utils.py
Comment (BestJuly) [inline megatron/core/context_parallel_layout/utils.py:1]: This file only hold one function. Can we move some common functions and make the structure more clear? E.g., this file renamed to utils.py and move some function such as also moving `_finalize_packed_seq_params` here.

File: megatron/core/transformer/transformer_config.py (hunk 5 above)
Comment (yuzhongw-nvidia) [reply megatron/core/transformer/transformer_config.py]: Very good question. The conversion does not support attn partial CG, either. But it is hard to assert this kind of case because we cannot know whether the conversion will happen in the config. Do you have any ideas?

**Review by BestJuly — APPROVED** [review]

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/582ea12997010601b061251478c0263830a4a68c

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6387/commits/63faa41548c891b9e00ee30c82835217a081d3f1

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/3bdaba274117be9b52d53ea714a67cbd9c812c36
